smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java

```diff
@@ -0,0 +1,262 @@
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns a value representing the StructureShape {@code shape} parameter that is
+     * serializable by {@code serializeDocument}.
+     *
+     * <p>The function signature for this body will have two parameters:
+     *   <ul>
+     *     <li>{@code input}: the type generated for the StructureShape shape parameter.</li>
+     *     <li>{@code context}: the SerdeContext.</li>
+     *   </ul>
+     *
+     * <p>For example, this function would generate the following:
+     *
+     * <pre>{@code
+     *   let bodyParams: any = {}
+     *   if (input.fooValue !== undefined) {
+     *     bodyParams['fooValue'] = serializeAws_restJson1_1Foo(input.fooValue, context);
+     *   }
+     *   if (input.barValue !== undefined) {
+     *     bodyParams['barValue'] = input.barValue;
+     *   }
+     *   return bodyParams;
+     * }</pre>
+     *
+     * @param context The generation context.
+     * @param shape The structure shape being generated.
+     */
+    void serializeStructure(GenerationContext context, StructureShape shape);
+
+    /**
+     * Writes the code needed to serialize a union in the document of a request.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns a value representing the UnionShape {@code shape} parameter that is
+     * serializable by {@code serializeDocument}.
+     *
+     * <p>The function signature for this body will have two parameters:
+     *   <ul>
+     *     <li>{@code input}: the type generated for the UnionShape shape parameter.</li>
+     *     <li>{@code context}: the SerdeContext.</li>
+     *   </ul>
+     *
+     * <p>For example, this function would generate the following:
+     * <pre>{@code
+     *   return Field.visit(input, {
+     *     fooValue: value => serializeAws_restJson1_1Foo(value, context),
+     *     barValue: value => value,
+     *     _: value => value
+     *   });
+     * }</pre>
+     *
+     * @param context The generation context.
+     * @param shape The union shape being generated.
+     */
+    void serializeUnion(GenerationContext context, UnionShape shape);
+
+    /**
+     * Writes the code needed to serialize a collection in the document of a request.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns a value representing the CollectionShape {@code shape} parameter that is
+     * serializable by {@code serializeDocument}.
+     *
+     * <p>The function signature for this body will have two parameters:
+     *   <ul>
+     *     <li>{@code input}: the type generated for the CollectionShape shape parameter.</li>
+     *     <li>{@code context}: the SerdeContext.</li>
+     *   </ul>
+     *
+     * <p>For example, this function would generate the following:
+     *
+     * <pre>{@code
+     *   return (input || []).map(entry =>
```

Comment:
ideally, you don't need to indent the contents of these methods manually, right?